### PR TITLE
Workaround an asio bug on OSX, async connect and clean up warnings

### PR DESCRIPTION
**Pull request by Meisaka**
## Description

The asio fix is to set work tracking and use `poll()` instead of `run()`
Removed some std::move that the compiler flagged for RVO
Made the client's Connect asynchronous, it will also show the addresses it resolved.
Made client's asio variables static, since there can only be one.

## Motivation and Context

On OS X, asio was not calling handlers in the client code, result was a brick.
Also got rid of a bunch of warnings, they were bothering me.

## How Has This Been Tested?

Run the tests, server, and client on Mac OS X

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: tests/net-message_test.cpp [hunk 1]
```diff
@@ -8,11 +8,10 @@
 namespace tec {
 namespace networking {
 
-// clang-format off
-// Warning: these are picky about whitespace, the s has to go right on the end to make a literal
 #define _STRINGIFY(a) #a
-#define CASE_LABEL(t) case t: return _STRINGIFY(t)s
-// clang-format on
+#define CASE_LABEL(t) \
```
Comment (adam4813) [inline tests/net-message_test.cpp:12]: Intentional?

File: client/server-connection.cpp [hunk 2]
```diff
@@ -46,33 +50,42 @@ bool ServerConnection::Connect(std::string_view ip) {
...
-	asio::ip::tcp::no_delay option(true);
-	this->socket.set_option(option);
-
-	if (this->onConnect) {
-		this->onConnect();
+	tcp::endpoint server_endpoint;
+	for (auto& query_result : query_results) {
+		auto query_point = query_result.endpoint(); // convert it to an endpoint
+		auto query_address = query_point.address();
+		if (query_address.is_v4()) { // only accept v4 for now, TODO: make sure IPv6 works too
```
Comment (adam4813) [inline client/server-connection.cpp:65]: Can you add a github issue so this isn't lost in the code?

File: client/server-connection.cpp [hunk 3]
```diff
@@ -84,6 +97,8 @@ void ServerConnection::Disconnect() {
 }
 
 void ServerConnection::Stop() {
+	io_work = asio::any_io_executor();
```
Comment (adam4813) [inline client/server-connection.cpp:100]: Not sure what's going on here. Is it grabbing a handle to do work after the close that might still be in the pipie?

**Review by adam4813 — APPROVED** [review]
Just had a few questions.

File: client/server-connection.cpp (hunk 3 above)
Comment (Meisaka) [reply client/server-connection.cpp:100]: I basically pulled the example from [The asio page on io_contexts](https://think-async.com/Asio/asio-1.18.1/doc/asio/reference/io_context.html)
and from what I understand, this "resets" the work tracking object, telling asio that's it's done.

File: tests/net-message_test.cpp (hunk 1 above)
Comment (Meisaka) [reply tests/net-message_test.cpp:12]: I removed the "s" from the end intentionally, which is what the format guards were only there for, the old compiler on my Mac wasn't having it.
Then without them, a later run of clang-format insisted on adding the like break: `\`

**Pull request merged**